x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/inference/aggs/InferencePipelineAggregationBuilder.java

```diff
@@ -0,0 +1,236 @@
+        this.modelLoadingService = modelLoadingService;
+        this.bucketPathMap = bucketsPath;
+    }
+
+    public InferencePipelineAggregationBuilder(StreamInput in, SetOnce<ModelLoadingService> modelLoadingService) throws IOException {
+        super(in, NAME);
+        modelId = in.readString();
+        bucketPathMap = in.readMap(StreamInput::readString, StreamInput::readString);
+        inferenceConfig = in.readOptionalNamedWriteable(InferenceConfigUpdate.class);
+        gapPolicy = BucketHelpers.GapPolicy.readFrom(in);
+        this.modelLoadingService = modelLoadingService;
+    }
+
+    void setModelId(String modelId) {
+        this.modelId = modelId;
+    }
+
+    void setInferenceConfig(InferenceConfigUpdate inferenceConfig) {
+        this.inferenceConfig = inferenceConfig;
+    }
+
+    void setGapPolicy(BucketHelpers.GapPolicy gapPolicy) {
+        this.gapPolicy = gapPolicy;
+    }
+
+    @Override
+    protected void validate(ValidationContext context) {
+        context.validateHasParent(NAME, name);
+        if (modelId == null) {
+            context.addValidationError("Model Id must be set");
+        }
+        if (gapPolicy != BucketHelpers.GapPolicy.SKIP) {
+            context.addValidationError("gap policy [" + gapPolicy + "] in not valid for [" + NAME + "] aggregation");
+        }
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeString(modelId);
+        out.writeMap(bucketPathMap, StreamOutput::writeString, StreamOutput::writeString);
+        out.writeOptionalNamedWriteable(inferenceConfig);
+        gapPolicy.writeTo(out);
+    }
+
+    @Override
+    protected PipelineAggregator createInternal(Map<String, Object> metaData) {
+
+        SetOnce<Model> model = new SetOnce<>();
+        SetOnce<Exception> error = new SetOnce<>();
+        CountDownLatch latch = new CountDownLatch(1);
+        ActionListener<Model> listener = new LatchedActionListener<>(
+            ActionListener.wrap(model::set, error::set), latch);
+
+        modelLoadingService.get().getModelForSearch(modelId, listener);
+        try {
+            // TODO Avoid the blocking wait
+            latch.await();
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException("Inference aggregation interrupted loading model", e);
+        }
+
+        Exception e = error.get();
+        if (e != null) {
+            if (e instanceof RuntimeException) {
+                throw (RuntimeException)e;
+            } else {
+                throw new RuntimeException(error.get());
+            }
+        }
+
+        InferenceConfigUpdate update = adaptForAggregation(inferenceConfig);
+
+        return new InferencePipelineAggregator(name, bucketPathMap, metaData, update, model.get());
+    }
+
+    static InferenceConfigUpdate adaptForAggregation(InferenceConfigUpdate originalUpdate) {
```

Comment:
I don't know what any of this is doing :), but I wanted to mention that by time this is called a lot of work will have been done on the data nodes.  So if any of this validation can be done earlier (parsing, etc) it would save some work.

Reply:
Very good point I've moved these checks to the `validate()` method